framework/13-inch/common/audio.nix

```diff
@@ -0,0 +1,368 @@
+            - Equalizer
+            - Slight compression
+
+          This option has been optimised for the Framework Laptop 13 AMD 7040 series, but should work on all models.
+
+          Before applying, ensure the speakers are set to 100%,
+          because the volumes compound and the raw speaker device will be hidden by default.
+
+          In some cases, the added bass will vibrate the keyboard cable leading to a rattling sound,
+          a piece of foam can be used to mitigate this.
+        '';
+      };
+
+      hideRawDevice = lib.mkOption {
+        type = lib.types.bool;
+        default = true;
+        description = ''
+          Hide the raw speaker device.
+          This option is enabled by default, because keeping the raw speaker device can lead to volume conflicts.
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enabled (let
+    outputName = "alsa_output.pci-0000_c1_00.6.analog-stereo";
+    prettyName = "Framework Speakers";
+
+    # These are pre-made decibel to linear value conversions, since Nix doesn't have pow().
+    # Use the formula `10 ** (db / 20)` to calculate.
+    db = {
+      "-18.1" = 0.1244514611771385;
+      "-5.48" = 0.5321082592667942;
+      "-4.76" = 0.5780960474057181;
+      "8.1" = 2.5409727055493048;
+      "-36" = 1.5848931924611134e-2;
+    };
+
+    json = pkgs.formats.json { };
+
+    # The filter chain, heavily inspired by the asahi-audio project: https://github.com/AsahiLinux/asahi-audio
+    filter-chain = json.generate "filter-chain.json" {
+      "node.description" = prettyName;
+      "media.name" = prettyName;
+      "filter.graph" = {
+        nodes = [
+          # Psychoacoustic bass extension,
+          # it creates harmonics of the missing bass to fool our ears into hearing it.
+          {
+            type = "lv2";
+            plugin = "https://chadmed.au/bankstown";
+            name = "bassex";
+            control = {
+              bypass = 0;
+              amt = 1.2;
+              sat_second = 1.3;
+              sat_third = 2.5;
+              blend = 1.0;
+              ceil = 200.0;
+              floor = 20.0;
+            };
+          }
+          # Loudness compensation,
+          # it ensures that the sound profile stays consistent across different volumes.
+          {
+            type = "lv2";
+            plugin = "http://lsp-plug.in/plugins/lv2/loud_comp_stereo";
+            name = "el";
+            control = {
+              enabled = 1;
+              input = 1.0;
+              fft = 4;
+            };
+          }
+          # 8-band equalizer,
+          # it tries to lessen frequencies where the laptop might resonate,
+          # and tries to make the frequency curve more pleasing;
+          # this is the "Lappy McTopface" profile (https://github.com/ceiphr/ee-framework-presets)
+          # further tuned for the Framework Laptop 13 AMD 7040 series
+          # and might need some tuning on other models.
```

Comment:
How would one tune this?

Reply:
The way I tuned it is that I used EasyEffects with plugin GUIs enabled and added the equalizer there. I tweaked the knobs until it sounded good and exported the options to a txt and used those numbers here.

Reply:
Do you know how can get the underlying filter names in easy effects that you are using here?

Reply:
Good enough.

Reply:
This is a bit complicated, but here's how I did it:
1. Open EasyEffect and set up the equalizer how you like it
2. Click "Show native window", this will open the LV2 plugin window
3. On the native window, there's a "Save" button which lets you save the equalizer state
4. I then trimmed the unused bands (EasyEffects uses a 32-band LR equalizer)


![EasyEffect](https://github.com/user-attachments/assets/849b9504-80c8-4e63-b2e7-d44b4babe55e)
![Native plugin window](https://github.com/user-attachments/assets/4e7c0cc0-5b20-4726-90cd-aea2d0d44562)


To get the plugin names, you can do this:
1. Run `systemctl --user cat wireplumber` and take note of the `LV2_PATH` env var
2. In the folder that `LV2_PATH` is set to, you'll find all the plugin packages specified in `passthru.requiredLv2Packages`.
3. Change into the package's folder and grep for the name.
4. Once you find the plugin's file, you can get the name from there.
For example, for me `LV2_PATH` is `/nix/store/c0pza0jwsa78h87lylr2jhi4aw3l8llm-wireplumber-lv2-plugins/lib/lv2`, I can then `cd` into `/nix/store/c0pza0jwsa78h87lylr2jhi4aw3l8llm-wireplumber-lv2-plugins/lib/lv2/lsp-plugins.lv2` and run `grep -Ri parametric .`.